File: Backend.JAVA/src/main/java/com/example/demo/controller/UserController.java
package com.example.demo.controller;

import java.time.LocalTime;

import org.slf4j.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;
import java.util.HashMap;
import java.util.Map;

import com.example.demo.Entity.User;
import com.example.demo.Service.JwtService;
import com.example.demo.Service.UserService;

@RestController
@CrossOrigin(origins = "http://localhost:3000")
public class UserController {

	@Autowired
	private UserService userService;

	@Autowired
	private AuthenticationManager authManager;

	@Autowired
	private JwtService jwtService;

	@Autowired
	private com.example.demo.Repository.CustomerRepository customerRepository;

	private static final Logger logger = LoggerFactory.getLogger(UserController.class);

	@PostMapping("/register")
	public ResponseEntity<User> register(@RequestBody User user) {
		// Enforce CUSTOMER role for public registration
		user.setRole(com.example.demo.Entity.Role.CUSTOMER);
		User user2 = userService.addUser(user);
		if (user2 == null) {
			return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(null);
		}
		return ResponseEntity.status(HttpStatus.CREATED).body(user);
	}

	@PostMapping(value = "/login")
	public ResponseEntity<?> login(@RequestBody User user) {
		try {
			Authentication authentication = authManager.authenticate(
					new UsernamePasswordAuthenticationToken(user.getUsername(), user.getPassword()));

			if (authentication.isAuthenticated()) {
				logger.info("Signed In Successfully " + user.getUsername());
				String jwt = jwtService.generateToken(user.getUsername());

				// Fetch full user details
				User fullUser = userService.getUserByUsername(user.getUsername());

				Map<String, Object> response = new HashMap<>();
				response.put("token", jwt);
				response.put("role", fullUser.getRole());
				response.put("userId", fullUser.getId());
				response.put("email", fullUser.getEmail());
				response.put("username", fullUser.getUsername());
				if (fullUser.getHub() != null) {
					response.put("hubId", fullUser.getHub().getHubId());
				}

				com.example.demo.Entity.CustomerMaster customer = customerRepository
						.findByEmailIgnoreCase(fullUser.getEmail());
				if (customer != null) {
					response.put("customerId", customer.getCustId());
				} else {
					// Fallback using userId if no customer record found, though this is risky
					response.put("customerId", fullUser.getId());
				}

				return ResponseEntity.ok(response);
			} else {
				return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Authentication failed");
			}
		} catch (Exception e) {
			logger.error("Sign In Error", e);
			return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Login Error: " + e.getMessage());
		}
	}

	@Autowired
	private com.example.demo.Service.GoogleAuthService googleAuthService;

	@PostMapping("/api/v1/auth/google")
	public ResponseEntity<?> googleLogin(@RequestBody Map<String, String> request) {
		try {
			String token = request.get("token");
			String jwt = googleAuthService.verifyGoogleTokenAndGetJwt(token);

			String username = jwtService.extractUserName(jwt);
			User fullUser = userService.getUserByUsername(username);

			Map<String, Object> response = new HashMap<>();
			response.put("token", jwt);
			response.put("role", fullUser.getRole());
			response.put("userId", fullUser.getId());
			response.put("email", fullUser.getEmail());
			response.put("username", fullUser.getUsername());
			if (fullUser.getHub() != null) {
				response.put("hubId", fullUser.getHub().getHubId());
			}

			com.example.demo.Entity.CustomerMaster customer = customerRepository
					.findByEmailIgnoreCase(fullUser.getEmail());
			if (customer != null) {
				response.put("customerId", customer.getCustId());
			} else {
				response.put("customerId", fullUser.getId());
			}

			return ResponseEntity.ok(response);
		} catch (Exception e) {
			logger.error("Google Sign In Error", e);
			return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Google Error: " + e.getMessage());
		}
	}

	@Autowired
	private com.example.demo.Service.EmailService emailService;

	@PostMapping("/forgot-password")
	public ResponseEntity<?> forgotPassword(@RequestBody Map<String, String> request) {
		String email = request.get("email");
		String token = userService.generateResetToken(email);
		if (token != null) {
			emailService.sendPasswordResetEmail(email, token);
		}
		// Always return OK for security reasons even if email is not found
		return ResponseEntity.ok("If an account exists with this email, you will receive a reset link shortly.");
	}

	@PostMapping("/reset-password")
	public ResponseEntity<?> resetPassword(@RequestBody Map<String, String> request) {
		String token = request.get("token");
		String newPassword = request.get("password");
		boolean success = userService.resetPassword(token, newPassword);
		if (success) {
			return ResponseEntity.ok("Password has been reset successfully.");
		} else {
			return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Invalid or expired reset token.");
		}
	}

}

File: Backend.NET/Controllers/UserController.cs
using Microsoft.AspNetCore.Mvc;
using FleetManagementSystem.Api.Models;
using FleetManagementSystem.Api.Services;
using FleetManagementSystem.Api.Data;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace FleetManagementSystem.Api.Controllers;

[ApiController]
public class UserController : ControllerBase
{
    private readonly IUserService _userService;
    private readonly IJwtService _jwtService;
    private readonly ApplicationDbContext _context; // Replacing CustomerRepository
    private readonly IGoogleAuthService _googleAuthService;
    private readonly IEmailService _emailService;
    private readonly ILogger<UserController> _logger;

    public UserController(IUserService userService, IJwtService jwtService, ApplicationDbContext context, IGoogleAuthService googleAuthService, IEmailService emailService, ILogger<UserController> logger)
    {
        _userService = userService;
        _jwtService = jwtService;
        _context = context;
        _googleAuthService = googleAuthService;
        _emailService = emailService;
        _logger = logger;
    }

    [HttpPost("register")]
    public async Task<IActionResult> Register([FromBody] RegisterRequest request)
    {
        var user = new User
        {
            Username = request.Username,
            Password = request.Password,
            Email = request.Email,
            Role = Role.CUSTOMER.ToString()
        };

        var createdUser = await _userService.AddUserAsync(user);
        if (createdUser == null)
        {
            return BadRequest("User creation failed. Username or Email might already exist.");
        }
        return Created("", createdUser); 
    }

    [HttpPost("login")]
    public async Task<IActionResult> Login([FromBody] LoginRequest loginRequest)
    {
        _logger.LogInformation("Login attempt for user: {Username}", loginRequest.Username);
        try
        {
            if (await _userService.ValidateCredentialsAsync(loginRequest.Username, loginRequest.Password))
            {
                // Authenticated
                 var jwt = _jwtService.GenerateToken(loginRequest.Username);
                 var fullUser = await _userService.GetUserByUsernameAsync(loginRequest.Username);

                 var response = new Dictionary<string, object>
                 {
                     { "token", jwt },
                     { "role", fullUser.Role },
                     { "userId", fullUser.Id },
                     { "email", fullUser.Email },
                     { "username", fullUser.Username }
                 };

                 if (fullUser.HubId.HasValue)
                 {
                     response["hubId"] = fullUser.HubId;
                 }

                 var customer = await _context.Customers.FirstOrDefaultAsync(c => c.Email == fullUser.Email);
                 if (customer != null)
                 {
                     response["customerId"] = customer.CustId;
                 }
                 else
                 {
                     response["customerId"] = fullUser.Id;
                 }

                 return Ok(response);
            }
            else
            {
                _logger.LogWarning("Invalid login attempt for user: {Username}", loginRequest.Username);
                return Unauthorized("Authentication failed");
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error during login for user: {Username}", loginRequest.Username);
            return Unauthorized("Invalid Creds");
        }
    }

    [HttpPost("api/v1/auth/google")]
    public async Task<IActionResult> GoogleLogin([FromBody] Dictionary<string, string> request)
    {
        try
        {
            if (request.TryGetValue("token", out var token))
            {
                var authResponse = await _googleAuthService.VerifyGoogleTokenAndGetJwtAsync(token);
                return Ok(authResponse);
            }
             return Unauthorized("Invalid Google Token");
        }
        catch (Exception ex)
        {
            return StatusCode(500, "Internal Server Error: " + ex.Message + " | " + ex.InnerException?.Message);
        }
    }

    [HttpPost("forgot-password")]
    public async Task<IActionResult> ForgotPassword([FromBody] Dictionary<string, string> request)
    {
        if (request.TryGetValue("email", out var email))
        {
            var token = await _userService.GenerateResetTokenAsync(email);
            if (token != null)
            {
                await _emailService.SendPasswordResetEmailAsync(email, token);
            }
        }
        return Ok("If an account exists with this email, you will receive a reset link shortly.");
    }

    [HttpPost("reset-password")]
    public async Task<IActionResult> ResetPassword([FromBody] Dictionary<string, string> request)
    {
        var token = request.TryGetValue("token", out var t) ? t : null;
        var password = request.TryGetValue("password", out var p) ? p : null;

        if (token != null && password != null && await _userService.ResetPasswordAsync(token, password))
        {
            return Ok("Password has been reset successfully.");
        }
        return BadRequest("Invalid or expired reset token.");
    }
}
